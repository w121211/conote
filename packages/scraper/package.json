{
  "name": "@conote/scraper",
  "version": "0.1.0",
  "dependencies": {
    "cheerio": "^1.0.0-rc.10",
    "got": "^11.8.2",
    "lodash": "^4.17.21",
    "memoize-one": "^5.2.1"
  },
  "devDependencies": {
    "@types/jest": "^27.4.0",
    "@types/lodash": "^4.14.177",
    "@types/node": "^18.0.3",
    "@typescript-eslint/eslint-plugin": "^5.5.0",
    "@typescript-eslint/parser": "^5.5.0",
    "eslint": "^8.4.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^3.3.1",
    "jest": "^27.4.3",
    "prettier": "^2.5.0",
    "ts-jest": "^27.0.7",
    "ts-node": "^10.7.0",
    "typescript": "^4.7.4"
  },
  "engines": {
    "node": ">=10.0.0"
  },
  "license": "MIT",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "scripts": {
    "build": "yarn run clean && yarn run build-ts",
    "build-ts": "tsc --build",
    "clean": "rm -rf dist && rm -rf reports",
    "test": "jest --verbose --watch"
  },
  "types": "./dist/index.d.ts"
}
