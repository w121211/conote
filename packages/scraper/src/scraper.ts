/**
 * Code mostly borrow from metascraper, @see
 */
import cheerio from 'cheerio'
import { fromPairs, isNull, isString } from 'lodash'
import memoizeOne from 'memoize-one'
import { Rule, RuleMap, RuleParams, Scraper, ScrapeResult, ScraperPack } from './types'

const hasValue = (value: unknown | unknown[]) => {
  if (Array.isArray(value)) {
    return value.length > 0
  }
  return value !== undefined && !Number.isNaN(value)
}

const normalizeValue = (value: unknown | unknown[]) => (hasValue(value) ? value : null)

const isKeyOfRuleMap = (k: string): k is keyof RuleMap => {
  return ['keywords', 'tickers'].includes(k)
}

const createMatchUrlFn = (domain: string) => {
  return (url: string): boolean => {
    return url.includes(domain)
  }
}
const withMatchUrl = (rule: Rule, matchUrl: (url: string) => boolean): Rule => {
  return ({ url, ...rest }: RuleParams) => {
    if (matchUrl(url)) {
      return rule({ url, ...rest })
    }
    return undefined
  }
}

const mergePacks = (packs: ScraperPack[]): RuleMap => {
  const merged: RuleMap = {}
  for (const pack of packs) {
    const matchUrlFn = pack.matchUrl ? memoizeOne(createMatchUrlFn(pack.matchUrl)) : undefined
    Object.entries(pack.ruleMap).forEach(([k, v]) => {
      if (isKeyOfRuleMap(k)) {
        const rules = matchUrlFn ? v.map((e): Rule => withMatchUrl(e, matchUrlFn)) : v
        merged[k] = [...(merged[k] ?? []), ...rules]
      } else {
        throw `isKeyOfRuleMap(k) === false, rule map key is not defined in ScrapeResult, ${k}`
      }
    })
  }
  return merged
}

const tryRule = async (rules: Rule[], params: RuleParams) => {
  for (const rule of rules) {
    // eslint-disable-next-line no-await-in-loop
    const value = await Promise.resolve(rule(params)) // fake promise
    if (hasValue(value)) {
      return value
    }
  }
  return undefined
}

const scrape = async ({ ruleMap, ...ruleParams }: { ruleMap: RuleMap } & RuleParams): Promise<ScrapeResult> => {
  const normalize = (v: string | string[] | null | undefined): string[] | undefined => {
    if (isString(v)) {
      return v.split(',').map(e => e.trim()) // try to parse
    }
    if (isNull(v)) {
      return undefined
    }
    return v
  }

  const promises = Object.entries(ruleMap).map(
    async ([k, v]): Promise<[string, string | string[] | null | undefined]> => {
      const value = await tryRule(v, ruleParams)
      // return [k, normalizeValue(value)]
      return [k, value]
    },
  )

  const data = fromPairs(await Promise.all(promises))
  const result: ScrapeResult = {
    keywords: normalize(data.keywords),
    tickers: normalize(data.tickers),
  }
  return result
}

export const createScraper = (packs: ScraperPack[]): Scraper => {
  const merged = mergePacks(packs)
  const scraper: Scraper = async ({ url, html }) => {
    return scrape({
      ruleMap: merged,
      url,
      $: cheerio.load(html),
      // ...kwargs,
    })
  }
  return scraper
}
