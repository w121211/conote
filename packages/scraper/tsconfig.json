{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "dist",
    // "composite": true,
    "declaration": true
  },
  "include": ["src"],
  "exclude": ["**/__deprecated__", "node_modules"]
}
