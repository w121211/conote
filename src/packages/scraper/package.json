{
  "name": "@conote/scraper",
  "version": "0.1.0",
  "license": "MIT",
  "scripts": {
    "clean": "rm -rf dist && rm -rf reports",
    "build-ts": "tsc --build",
    "build": "yarn run clean && yarn run build-ts",
    "test": "jest --verbose"
  },
  "main": "./dist/src/index.js",
  "module": "./dist/src/index.js",
  "types": "./dist/src/index.d.ts",
  "devDependencies": {
    "@types/jest": "^26.0.20",
    "@types/lodash": "^4.14.175",
    "@types/node": "^13.13.21",
    "@typescript-eslint/eslint-plugin": "^4.16.1",
    "@typescript-eslint/parser": "^4.16.1",
    "eslint": "^6.8.0",
    "eslint-config-prettier": "^7.2.0",
    "eslint-plugin-prettier": "^3.3.1",
    "jest": "^25.5.4",
    "prettier": "^2.2.1",
    "ts-jest": "^26.5.3",
    "ts-node": "^8.10.2",
    "typescript": "^4.2.2"
  },
  "dependencies": {
    "cheerio": "^1.0.0-rc.10",
    "got": "^11.8.5",
    "lodash": "^4.17.21",
    "memoize-one": "^5.2.1"
  },
  "engines": {
    "node": ">=10.0.0"
  }
}
