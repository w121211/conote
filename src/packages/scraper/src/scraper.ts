import { decodeHTML } from 'entities'

import cheerio, { CheerioAPI } from 'cheerio'
import {
  chain,
  concat,
  eq,
  flow,
  forEach,
  fromPairs,
  get,
  has,
  includes,
  invoke,
  isArray,
  isBoolean,
  isDate,
  isEmpty,
  isNumber,
  isString,
  lte,
  map,
  replace,
  set,
  size,
  toLower,
  toPairs,
  toString,
} from 'lodash'
import memoizeOne from 'memoize-one'

// interface Metadata {
//   author: string
//   date: string
//   description: string
//   image: string
//   publisher: string
//   title: string
//   url: string
// }

export type RuleProps = {
  url: string
  $: CheerioAPI
}

export type Rule = (props: RuleProps) => string[] | string | null | undefined

type RuleSet = Record<string, Rule[]>

type ScraperProps = {
  url: string
  html: string
  // ruleSet: RuleSet[]
}

type Scraper = (props: ScraperProps) => Promise<Record<string, string | string[] | undefined>>

const jsonld = memoizeOne(
  ($: CheerioAPI) =>
    $('script[type="application/ld+json"]')
      .map((i, e) => {
        try {
          return JSON.parse($(e).contents().text())
        } catch (err) {
          return undefined
        }
      })
      .get()
      .filter(Boolean),
  (newArgs, lastArgs) => newArgs[0].html() === lastArgs[0].html(),
)

const $jsonld = (propName: string) => ($: CheerioAPI) => {
  const collection = jsonld($)
  let value

  console.log(collection)

  collection.find(item => {
    value = get(item, propName)
    return !isEmpty(value) || isNumber(value) || isBoolean(value)
  })

  return isString(value) ? decodeHTML(value) : value
}

const hasValue = (value: unknown) => value !== undefined && !Number.isNaN(value)

const normalizeValue = (value: unknown) => (hasValue(value) ? value : null)

// const forEachRule = (collection, fn) => forEach(castArray(collection), fn)

// const loadRules = (rulesBundle: RuleSet[]) =>
//   chain(rulesBundle)
//     .reduce((acc, { test, ...rules }) => {
//       forEach(rules, (innerRules, propName) => {
//         if (test) forEachRule(innerRules, rule => (rule.test = test))
//         set(acc, propName, has(acc, propName) ? concat(acc[propName], innerRules) : concat(innerRules))
//         return acc
//       })
//       return acc
//     }, {})
//     .toPairs()
//     .value()

// const mergeRules = (rules, baseRules) =>
//   chain(rules)
//     .reduce((acc, { test, ...rules }) => {
//       forEach(rules, (innerRules, propName) => {
//         if (test) forEachRule(innerRules, rule => (rule.test = test))
//         // find the rules associated with `propName`
//         const index = findIndex(acc, item => first(item) === propName)
//         // if `propName` has more rule, add the new rule from the end
//         if (index !== -1) acc[index][1] = concat(innerRules, ...acc[index][1])
//         // otherwise, create an array of rules
//         else acc.push([propName, castArray(innerRules)])
//       })
//       return acc
//     }, cloneDeep(baseRules))
//     .value()

const findRule = async (rules: Rule[], props: RuleProps) => {
  const has = (value: unknown): boolean => value !== undefined && !Number.isNaN(value)
  // let index = 0
  // let value
  // do {
  //   const rule = rules[index++]
  //   const test = rule.test ?? truthyTest
  //   if (test(args)) value = await rule(kwargs)
  // } while (!has(value) && index < rules.length)
  // return value
  for (const rule of rules) {
    // if (check.test) {
    //   check.test(args)
    // }
    // eslint-disable-next-line no-await-in-loop
    const value = await Promise.resolve(rule(props)) // fake promise
    if (has(value)) {
      return value
    }
  }
  return undefined
}

async function getData({ ruleSet, ...props }: { ruleSet: RuleSet } & RuleProps) {
  const rulePairs = toPairs(ruleSet)
  const data = await Promise.all(
    map(rulePairs, async ([propName, rules]) => {
      const value = await findRule(rules, props)
      return [propName, normalizeValue(value)]
    }),
  )
  return fromPairs(data)
}

export function createScraper(ruleSets: RuleSet[]): Scraper {
  const merged = ruleSets[0]
  const scraper: Scraper = async ({ url, html }) => {
    return getData({
      url,
      $: cheerio.load(html),
      // rules: mergeRules(inlineRules, loadedRules),
      ruleSet: merged,
      // ...kwargs,
    })
  }
  return scraper
}
