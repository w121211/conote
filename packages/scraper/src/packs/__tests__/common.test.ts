import got from 'got'
import { createScraper } from '../../scraper'
import keywordsPack from '../keywords'
import tickersPack from '../tickers'

const testUrls = [
  'https://www.mobile01.com/topicdetail.php?f=793&t=6520838',
  'https://news.cnyes.com/news/id/4738961',
  'https://news.cnyes.com/news/id/4741929?exp=a',
]

it.each(testUrls)('keywords %s', async url => {
  const scraper = createScraper([keywordsPack])
  const { body: html, url: finalUrl } = await got(url)
  const { keywords } = await scraper({ url: finalUrl, html })
  expect(keywords).toMatchSnapshot()
})

it.each(testUrls)('tickers %s', async url => {
  const scraper = createScraper([tickersPack])
  const { body: html, url: finalUrl } = await got(url)
  const { tickers } = await scraper({ url: finalUrl, html })
  expect(tickers).toMatchSnapshot()
})
