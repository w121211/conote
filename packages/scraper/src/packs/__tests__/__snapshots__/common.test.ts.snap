// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keywords https://news.cnyes.com/news/id/4738961 1`] = `
Array [
  "本周大事",
  "台積電",
  "國安基金",
  "宏碁",
  "宏達電",
]
`;

exports[`keywords https://news.cnyes.com/news/id/4741929?exp=a 1`] = `
Array [
  "萬寶投顧",
  "楊惠宇",
  "台股",
]
`;

exports[`keywords https://www.mobile01.com/topicdetail.php?f=793&t=6520838 1`] = `
Array [
  "股票 / ETF",
  "理財",
]
`;

exports[`tickers https://news.cnyes.com/news/id/4738961 1`] = `undefined`;

exports[`tickers https://news.cnyes.com/news/id/4741929?exp=a 1`] = `
Array [
  "5299-TW",
  "8261-TW",
  "3317-TW",
  "6435-TW",
  "6138-TW",
  "6217-TW",
  "6510-TW",
  "6223-TW",
]
`;

exports[`tickers https://www.mobile01.com/topicdetail.php?f=793&t=6520838 1`] = `undefined`;
